services/src/handlers/mod.rs

```diff
@@ -3,40 +3,64 @@ use crate::error::Result;
 use crate::users::session::SessionId;
 use crate::users::userdb::UserDB;
 use crate::{contexts::Context, error::Error};
-use serde_json::json;
+use serde::{Deserialize, Serialize};
 use snafu::ResultExt;
 use std::error::Error as StdError;
 use std::str::FromStr;
-use warp::http::StatusCode;
-use warp::Filter;
-use warp::{Rejection, Reply};
+use warp::http::{Response, StatusCode};
+use warp::hyper::body::Bytes;
+use warp::reject::{InvalidQuery, MethodNotAllowed};
+use warp::{Filter, Rejection, Reply};
 
 pub mod projects;
 pub mod users;
 pub mod wfs;
 pub mod wms;
 pub mod workflows;
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ErrorResponse {
+    pub error: String,
+    pub message: String,
+}
+
+impl ErrorResponse {
+    pub fn assert(res: &Response<Bytes>, status: u16, error: &str, message: &str) {
+        assert_eq!(res.status(), status);
+
+        let body = std::str::from_utf8(&res.body()).unwrap();
+        assert_eq!(
+            serde_json::from_str::<ErrorResponse>(body).unwrap(),
+            ErrorResponse {
+                error: error.to_string(),
+                message: message.to_string(),
+            }
+        );
+    }
+}
+
 /// A handler for custom rejections
 pub async fn handle_rejection(err: Rejection) -> Result<impl Reply, Rejection> {
     let (code, error, message) = if let Some(e) = err.find::<Error>() {
         // custom errors
 
         // TODO: distinguish between client/server/temporary/permanent errors
-        if let error::Error::Authorization { source: e } = e {
-            let error_name: &'static str = e.as_ref().into();
-            (
+        match e {
+            error::Error::Authorization { source } => (
                 StatusCode::UNAUTHORIZED,
-                error_name.to_string(),
+                Into::<&str>::into(source.as_ref()).to_string(),
+                source.to_string(),
+            ),
+            error::Error::RegistrationFailed { reason } if reason == "E-mail already exists" => (
```

Comment:
instead of matching the reason to a string you should use a separate error variant

Reply:
505e492